ghw/libghw.c

```diff
@@ -628,24 +709,28 @@ ghw_read_type_bounds (struct ghw_handler *h, union ghw_type *base)
     default:
       fprintf (stderr, "ghw_read_type_bounds: unhandled kind %d\n",
 	       base->kind);
-      return NULL;
+      ghw_error_exit();
     }
 }
 
 int
 ghw_read_type (struct ghw_handler *h)
 {
   unsigned char hdr[8];
-  unsigned i;
+  uint32_t i;
 
   if (fread (hdr, sizeof (hdr), 1, h->stream) != 1)
     return -1;
 
   if (hdr[0] != 0 || hdr[1] != 0 || hdr[2] != 0 || hdr[3] != 0)
     return -1;
-  h->nbr_types = ghw_get_i32 (h, &hdr[4]);
-  h->types =
-    (union ghw_type **) malloc (h->nbr_types * sizeof (union ghw_type *));
+  h->nbr_types = (uint32_t) ghw_get_i32 (h, &hdr[4]);
+
+  // Allocate type storage, initialize it to 0 to catch invalid accesses.
+  uint64_t h_types_size = (uint64_t)h->nbr_types * sizeof (union ghw_type *);
```

Comment:
Check non negative result, use `calloc`

Reply:
Done. (I don't think the negative value check is actually necessary here)